### ISSUE TITLE
`id` should not be defined until the object has been created

### ISSUE THREAD

**Issue by volbap** [inaka/Jayme#22]
This is an architectural concern...

Currently:
`func create(entity: EntityType) -> Future<Void, ServerBackendError>`
hits
`POST /:name/:id`

(notice the existence of an `id` at that point)

There's an issue there and it's that, usually, it should be up to the server to generate the `id` of the object during its creation. Any object, therefore, should not have any associated `id` value until it's been actually created in the server and we have gotten a successful response from that process.

**Proposal:**
- Make the `create(...)` function hit `POST /:name` instead of `POST /:name/:id`.
- Replace the `Identifier` typealias in the `Identifiable` protocol so that it uses a custom `enum` (instead of a plain `String`), which would have 2 cases:

```
enum Identifier {
    case Undefined
    case Some(String)
}
```

or, let's just use an **optional**, which works exactly like that wrapper:
`typealias Identifier: String?`, although this option is not as clear as the explicit `enum` regarding intention.

Also, `create(...)` in `CRUDRepository` should not return `Void` in its `future`, but the created `User` with its corresponding `id` coming from the server; otherwise, there's no way to get the `id` back.

**Comment by volbap** [comment]
`typealias Identifier: String!` is another option too.

**Comment by volbap** [comment]
Using the explicit `enum` would carry the issue of having to perform pattern-matching every time we need to get an `id` from an object. Because of that, I would just throw away this option.

**Comment by volbap** [comment]
**Proposed solution:**
- Redefine the `Identifiable` protocol such that we let users choose which type they want to use for their identifiers.

``` swift
public protocol Identifiable {
    associatedtype IdentifierType: CustomStringConvertible
    var id: IdentifierType { get }
}
```
- The `IdentifierType` has to conform to `CustomStringConvertible` so that `CRUDRepository` can make up a `Path` by interpolating strings:

``` swift
private func pathForID(id: EntityType.IdentifierType) -> Path {
    return "\(self.name)/\(id)"
}
```
- This way, we do not tie users to have to use a particular type for their identifiers, as we've been doing up to now (when attaching `typealias Identifier = String`). Thus, we give them the freedom of choosing whatever identifier type is more convenient for them. What's more, with this architecture, each `EntityType` can define its own `IdentifierType`. This means a `Car` can have a `var id: String`, a `Bottle` can have a `var id: Int`, and so on.
- With this approach, we put aside the issue of defining who's responsible of creating id's (client vs. server), because we leave up to the users to define that. For instance, sometimes it happens that your entity needs to have an `id` while it's being processed in the server (say you need to display them on screen and for some reason you need to identify them univocally, then you do need a client-defined `id`), but afterwards, to keep consistency with the server, you will need for your entity to use the `id` that the server sends upon creation, overwriting the (non useful anymore) client-defined `id` that your entity had. In such a scenario, **users** can come up with a structure like this one:
  
  ``` swift
  enum Identifier {
  case Local(String)
  case Server(String)
  }
  
  extension Identifier: CustomStringConvertible {
  var description: String {
      switch self {
      case .Local(let id): return "local_\(id)"
      case .Server(let id): return id
      }
  }
  }
  ```
- By implementing this solution, the `:id` parameter would still be sent to the server in the `create(...)` method of `CRUDRepository`, but still, if that's not needed by the user, he can still implement its own `create(...)` and make it not send such value.
- Getting the entity `id` from the server will be possible when we solve issue #37.

**Comment by volbap** [comment]
Solved in PR #42 using latest proposed solution.

**Comment by volbap** [comment]
As @elbrujohalcon states, we can have a `POST` request containing an `id`, but that `id` would appear within the body of the request, not in the URL.
We need to change that. I'm going to use issue #63 for it.
